rest-server/templates/src/main/scala/com/actionml/templates/cb/CBEngine.scala

```diff
@@ -68,7 +75,8 @@ class CBEngine() extends Engine() with JsonParser {
   }
 
   /** Triggers parse, validation, and persistence of event encoded in the json */
-  def input(json: String, trainNow: Boolean = true): Validated[ValidateError, Boolean] = {
+  // Slava this needs to stay "input" why did you change the API?
+  def inputInternal(json: String, trainNow: Boolean = true): Validated[ValidateError, Boolean] = {
```

Comment:
why not call it `input`? Then have it call `super.input` something like below. That way the implementation can be put in `Engine` for all Templates to inherit.

```
super.input(json).andThen { p =>
  logger.trace("Got JSON body: " + json)
  // validation happens as the input goes to the dataset
  dataset.input(json).andThen(process(_)).map(_ => true)
}
```